### PR TITLE
*: reimplement dynamic config

### PR DESCRIPTION
**Pull request by rleungx**
### What is changed and how it works?
This PR removes the original implementation and adds new APIs to register and get addresses of other components. 

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: pkg/component/manager.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+		addresses = append(addresses, ca...)
+	}
+	return addresses
+}
+
+// GetAllComponentAddrs returns all components' addresses.
+func (c *Manager) GetAllComponentAddrs() map[string][]string {
+	c.RLock()
+	defer c.RUnlock()
+	return c.Addresses
```
Comment (HunDunDM) [inline pkg/component/manager.go:55]: Need to clone, otherwise there will be data race problems.

File: pkg/component/manager.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Register is used for registering a component with an address to PD.
+func (c *Manager) Register(component, addr string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	str := strings.Split(addr, ":")
+	if len(str) != 0 {
+		ip := net.ParseIP(str[0])
```
Comment (HunDunDM) [inline pkg/component/manager.go:77]: address is not necessarily ip? And should it come with http or https?

File: pkg/component/manager.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	for component, ca := range c.Addresses {
+		if contains(ca, addr) {
+			return component
+		}
+	}
+	return ""
+}
+
+// Register is used for registering a component with an address to PD.
+func (c *Manager) Register(component, addr string) error {
```
Comment (HunDunDM) [inline pkg/component/manager.go:71]: How to unregister?

File: tests/dashboard/service_test.go [hunk 4]
```diff
@@ -156,21 +144,6 @@ func (s *serverTestSuite) TestDashboard(c *C) {
 	s.checkServiceIsStarted(c, servers, leader)
 	c.Assert(leader.GetServer().GetScheduleOption().GetDashboardAddress(), Equals, dashboardAddress2)
 
-	// Changing dashboard address
```
Comment (HunDunDM) [inline tests/dashboard/service_test.go:159]: Do we have other ways to run this test?

Comment (rleungx) [reply tests/dashboard/service_test.go:159]: Any suggestion?

File: server/api/component_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	var output1 []string
+	err = readJSON(addr1, &output)
+	c.Assert(strings.Contains(err.Error(), "404"), IsTrue)
+	c.Assert(len(output1), Equals, 0)
+
+	// register 2 c1 and 1 c2
+	reqs := []map[string]string{
+		{"component": "c1", "addr": "127.0.0.1:1"},
+		{"component": "c1", "addr": "127.0.0.1:2"},
+		{"component": "c2", "addr": "127.0.0.1:3"},
```
Comment (overvenus) [inline server/api/component_test.go:63]: Could you add a test case to test domain address?

```suggestion
		{"component": "c2", "addr": "127.0.0.1:3"},
		{"component": "c3", "addr": "example.com"},
```

File: pkg/component/manager_test.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+func Test(t *testing.T) {
+	TestingT(t)
+}
+
+var _ = Suite(&testManagerSuite{})
+
+type testManagerSuite struct{}
+
+func (s *testManagerSuite) TestManager(c *C) {
```
Comment (HunDunDM) [inline pkg/component/manager_test.go:31]: Test the address that is not IP?

File: server/cluster/cluster.go [hunk 7]
```diff
@@ -621,6 +615,7 @@ func (c *RaftCluster) updateStoreStatusLocked(id uint64) {
 	c.core.UpdateStoreStatus(id, leaderCount, regionCount, pendingPeerCount, leaderRegionSize, regionSize)
 }
 
+//nolint
```
Comment (HunDunDM) [inline server/cluster/cluster.go:618]: why?

Comment (HunDunDM) [reply server/cluster/cluster.go:618]: We can limit the scope of `nolint`.

File: tests/dashboard/service_test.go (hunk 4 above)
Comment (HunDunDM) [reply tests/dashboard/service_test.go:160]: I will find a way to make up for this test later.

**Pull request merged**